source/common/router/router.cc

```diff
@@ -236,10 +249,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
     ENVOY_STREAM_LOG(debug, "unknown cluster '{}'", *callbacks_, route_entry_->clusterName());
 
     callbacks_->requestInfo().setResponseFlag(RequestInfo::ResponseFlag::NoRouteFound);
-    Http::HeaderMapPtr response_headers{new Http::HeaderMapImpl{
-        {Http::Headers::get().Status,
-         std::to_string(enumToInt(route_entry_->clusterNotFoundResponseCode()))}}};
-    callbacks_->encodeHeaders(std::move(response_headers), true);
+    sendLocalReply(route_entry_->clusterNotFoundResponseCode(), "");
```

Comment:
Could we make this a meaningful message?

Reply:
Tried it, made some tests fail (likely due to body in the response being somehow unexpected), so did not look at it further.